contracts/dapps/gauge/DappGauge.sol

```diff
@@ -0,0 +1,426 @@
+        require(gauges[pool].users[msg.sender].hasDeposited, "Caller not a depositor");
+        if (amount == 0) return;
+
+        _updateNutAcc();
+        _updatePoolAcc(pool);
+
+        address community = gauges[pool].community;
+        address factory = gauges[pool].factory;
+
+        amount = gauges[pool].users[msg.sender].amount > amount ? amount : gauges[pool].users[msg.sender].amount;
+
+        if(gauges[pool].users[msg.sender].amount > 0) {
+            // update user's reward include nut and ctoken
+            uint256 pendingNut = gauges[pool].users[msg.sender].amount.mul(userNutAcc).div(1e16).sub(gauges[pool].users[msg.sender].nutDebt);
+            uint256 pendingCToken = gauges[pool].users[msg.sender].amount.mul(gauges[pool].cTokenAcc).div(1e12).sub(gauges[pool].users[msg.sender].cTokenDebt);
+            gauges[pool].users[msg.sender].nutAvailable = gauges[pool].users[msg.sender].nutAvailable.add(pendingNut);
+            gauges[pool].users[msg.sender].cTokenAvailable = gauges[pool].users[msg.sender].cTokenAvailable.add(pendingCToken);
+        }
+        if (communityTotalStakedVeNut[community] > 0) {
+            // update community's reward only nut
+            uint256 commmunityPending = communityTotalStakedVeNut[community].mul(communityNutAcc).div(1e12).sub(communityDebt[community]);
+            communityAvailable[community] = communityAvailable[community].add(commmunityPending);
+        }
+        if (poolFactoryTotalStakedVeNut[factory] > 0) {
+            // update tool dev's reward only nut
+            uint256 poolFactoryPending = poolFactoryTotalStakedVeNut[factory].mul(poolFactoryNutAcc).div(1e12).sub(poolFactoryDebt[factory]);
+            poolFactoryAvailable[factory] = poolFactoryAvailable[factory].add(poolFactoryPending);
+        }
+
+        releaseERC20(veNUT, address(msg.sender), amount);
+
+        // update amount
+        gauges[pool].users[msg.sender].amount = gauges[pool].users[msg.sender].amount.sub(amount);
+        gauges[pool].totalStakedVeNut = gauges[pool].totalStakedVeNut.sub(amount);
+        communityTotalStakedVeNut[community] = communityTotalStakedVeNut[community].sub(amount);
+        poolFactoryTotalStakedVeNut[factory] = poolFactoryTotalStakedVeNut[factory].sub(amount);
+        totalVeNUTStaked = totalVeNUTStaked.sub(amount);
+
+        // update debt
+        gauges[pool].users[msg.sender].nutDebt = gauges[pool].users[msg.sender].amount.mul(userNutAcc).div(1e12);
+        gauges[pool].users[msg.sender].cTokenDebt = gauges[pool].users[msg.sender].amount.mul(gauges[pool].cTokenAcc).div(1e12);
+        communityDebt[community] = communityTotalStakedVeNut[community].mul(communityNutAcc).div(1e12);
+        poolFactoryDebt[factory] = poolFactoryTotalStakedVeNut[factory].mul(poolFactoryNutAcc).div(1e12);
+
+        emit Withdrawn(community, factory, pool, msg.sender, amount);
+    }
+
+    function userWithdrawReward(address pool) external nonReentrant {
```

Comment:
Maybe we can have a common private function so that user/community/dappFactory withdraw are no need impl the same logic many times

Reply:
Have considered this.But I think current version is more clear.